cmd/snap-preseed/preseed_linux.go

```diff
@@ -0,0 +1,172 @@
+// sure critical virtual filesystems (such as proc) are mounted. This is not meant to
+// be exhaustive check, but one that prevents running the tool against a wrong directory
+// by an accident, which would lead to hard to understand errors from snapd in preseed
+// mode.
+func checkChroot(preseedChroot string) error {
+	exists, isDir, err := osutil.DirExists(preseedChroot)
+	if err != nil {
+		return fmt.Errorf("cannot verify target chroot directory %s: %v", preseedChroot, err)
+	}
+	if !exists || !isDir {
+		return fmt.Errorf("target chroot directory %s doesn't exist or is not a directory", preseedChroot)
+	}
+
+	// sanity checks of the critical mountpoints inside chroot directory
+	for _, p := range []string{"/sys/kernel/security/apparmor", "/proc/self", "/dev/mem"} {
+		path := filepath.Join(preseedChroot, p)
+		if exists := osutil.FileExists(path); !exists {
+			return fmt.Errorf("target chroot directory validation error: %s doesn't exist", path)
+		}
+	}
+
+	return nil
+}
+
+var systemSnapFromSeeds = func(rootDir string) (string, error) {
+	seedDir := filepath.Join(dirs.SnapSeedDirUnder(rootDir))
+	seed, err := seed.Open(seedDir, "")
+	if err != nil {
+		return "", err
+	}
+
+	if err := seed.LoadAssertions(nil, nil); err != nil {
+		return "", err
+	}
+
+	tm := timings.New(nil)
+	if err := seed.LoadMeta(tm); err != nil {
+		return "", err
+	}
+
+	var coreSnapPath string
+	ess := seed.EssentialSnaps()
+	// TODO: handle core18, snapd snap.
+	for _, snap := range ess {
+		if snap.SnapName() == "core" {
+			coreSnapPath = snap.Path
+			break
+		}
+	}
+	if coreSnapPath == "" {
+		return "", fmt.Errorf("core snap not found")
+	}
+
+	return coreSnapPath, nil
+}
+
+func prepareChroot(preseedChroot string) (func(), error) {
+	if err := syscallChroot(preseedChroot); err != nil {
+		return nil, fmt.Errorf("cannot chroot into %s: %v", preseedChroot, err)
+	}
+
+	// GlobalRootDir is now relative to chroot env
+	rootDir := dirs.GlobalRootDir
+	coreSnapPath, err := systemSnapFromSeeds(rootDir)
+	if err != nil {
+		return nil, err
+	}
+
+	// create mountpoint for core/snapd
+	where := filepath.Join(rootDir, mountPath)
+	if err := os.MkdirAll(where, 0755); err != nil {
+		return nil, err
+	}
+
+	removeMountpoint := func() {
+		for i := 0; i < 5; i++ {
+			err := os.Remove(where)
+			if err != nil {
+				fmt.Fprintf(Stderr, "%v", err)
+			} else {
+				return
+			}
+			time.Sleep(time.Second)
+		}
+	}
+
+	cmd := exec.Command("mount", "-t", "squashfs", coreSnapPath, where)
+	if err := cmd.Run(); err != nil {
+		removeMountpoint()
+		return nil, fmt.Errorf("cannot mount %s at %s in preseed mode: %v ", coreSnapPath, where, err)
+	}
+
+	// TODO: check snapd version
+
+	unmount := func() {
+		fmt.Fprintf(Stdout, "unmounting: %s\n", mountPath)
+		cmd := exec.Command("umount", mountPath)
+		if err := cmd.Run(); err != nil {
+			fmt.Fprintf(Stderr, "%v", err)
+		}
+	}
+
+	return func() {
+		unmount()
```

Comment:
Should this exit the chroot and reset dirs.RootDir?

For exiting the chroot, you'll probably need to grab a file descriptor to `"/"` before chrooting and `f.Chdir()` to it in the cleanup.

Reply:
I don't think we should do that. This is a standalone utility and doing chroot work is its last operation. The cleanup functions are meant to clean up any artifacts from the filesystem.